### DISCUSSION TITLE
SPC Files from Secret of Evermore seem to be unsupported...

### DISCUSSION THREAD

**Discussion by dreimer1986** [vgmtrans/vgmtrans#934] [Music formats]
**Is your feature request related to a problem? Please describe.**

I look for ways to optimize the environmental tracks for the Secret of Evermore MSU patch. Problem is that I can't just use a similar environmental PCM file as the old ones are just burnt into my brain and I always miss something in them. So I wanted to load the SPC files, convert them to MIDI and try my luck from this side. Problem... VGMtrans does not detect these files as it seems. It's no AKAO format, that for sure...

**Describe the solution you'd like**

Well, I think added support for these files would be nice.

**Additional context**

Files can be found here: https://fi.zophar.net/soundfiles/nintendo-snes-spc/secret-of-evermore/Secret%20of%20Evermore%20%28EMU%29.zophar.zip

Some analyze efforts: https://www.ff6hacking.com/forums/thread-3638.html

**Comment by KungFuFurby** [comment]
The sound driver used by this game is by Sculptured Software. The US version's sound driver build is completely unique, but the European version's sound driver build is shared by WWF WrestleMania: The Arcade game.

This sound driver has multiple wrappers to deal with before we get to the internal VCMD format, and the VCMD format itself, if I remember correctly, uses relative notes, rather than absolute notes. My documentation is incomplete, though, particularly because I try to start from the earliest version possible.

**Comment by dreimer1986** [comment]
Almost thought that this variant might be a bit special. It's the only game that can claim to use half of it's tracks for ambience of environment and not music. Thanks for the interest in adding support if possible :D

**Comment by KungFuFurby** [comment]
Your game's pretty far down the list versioning-wise I'll note... the games that have the earliest versions, on the other hand, are Simpsons: Bart's Nightmare/Simpsons: Bart no Fushigi na Yume no Daibouken, Jack Nicklaus Golf, Monopoly, and WWF Super WrestleMania. The first one I'm not sure if it's the earliest period, however I did catch some noticeable memory location-related differences.

Also, this sound driver is notorious for being able to switch sample IDs on the fly through its own internal envelope support if I recall correctly. For added fun, it calibrates the GAIN values based on the next volume tick in a volume envelope (and it's not direct gain, either: it's a full blown volume envelope to ensure smooth transitions from one volume value to the next). Not only that, but it's among the earliest to also use pitch modulation (AKAO is the earliest that actually uses it that I know of, with Final Fantasy 4/2 being the first). On even later variants, of which Secret of Evermore is one of them, the sample directory is used in a non-standard way: namely, the sample pointers themselves are swapped out on a per-channel basis for the most part (a ninth slot is used as a backup of sorts). The actual pointers to the corresponding samples are stored further down, which at least reduces complications, but still means you're dealing with a non-standard utilization of the sample directory.

**Comment by LucianoTheWindowsFan** [comment]
> Your game's pretty far down the list versioning-wise I'll note... the games that have the earliest versions, on the other hand, are Simpsons: Bart's Nightmare/Simpsons: Bart no Fushigi na Yume no Daibouken, Jack Nicklaus Golf, Monopoly, and WWF Super WrestleMania. The first one I'm not sure if it's the earliest period, however I did catch some noticeable memory location-related differences.
> 
> Also, this sound driver is notorious for being able to switch sample IDs on the fly through its own internal envelope support if I recall correctly. For added fun, it calibrates the GAIN values based on the next volume tick in a volume envelope (and it's not direct gain, either: it's a full blown volume envelope to ensure smooth transitions from one volume value to the next). Not only that, but it's among the earliest to also use pitch modulation (AKAO is the earliest that actually uses it that I know of, with Final Fantasy 4/2 being the first). On even later variants, of which Secret of Evermore is one of them, the sample directory is used in a non-standard way: namely, the sample pointers themselves are swapped out on a per-channel basis for the most part (a ninth slot is used as a backup of sorts). The actual pointers to the corresponding samples are stored further down, which at least reduces complications, but still means you're dealing with a non-standard utilization of the sample directory.

IIRC the "sample pointers themselves swapped out on a per-channel basis" thingy was also used on Factor 5's sound driver.

**Comment by KungFuFurby** [comment]
I've also seen it in Frantic Flea.

**Comment by mikelow** [comment]
> Also, this sound driver is notorious for being able to switch sample IDs on the fly through its own internal envelope support if I recall correctly. For added fun, it calibrates the GAIN values based on the next volume tick in a volume envelope (and it's not direct gain, either: it's a full blown volume envelope to ensure smooth transitions from one volume value to the next). Not only that, but it's among the earliest to also use pitch modulation (AKAO is the earliest that actually uses it that I know of, with Final Fantasy 4/2 being the first). On even later variants, of which Secret of Evermore is one of them, the sample directory is used in a non-standard way: namely, the sample pointers themselves are swapped out on a per-channel basis for the most part (a ninth slot is used as a backup of sorts). The actual pointers to the corresponding samples are stored further down, which at least reduces complications, but still means you're dealing with a non-standard utilization of the sample directory.

@KungFuFurby You are a fount of knowledge. This format sounds like a pain in the ass.

I'm converting this to a discussion to keep unsupported formats out of the Issues list. Good info here.
